doc/decisions/0_drafts/keyname.md

```diff
@@ -0,0 +1,136 @@
+# Namespace and Name of Keys
+
+## Problem
+
+A `Key` in Elektra is identified by its name, which consists of a namespace and a number of name parts.
+There are two common representations for the name: escaped and unescaped.
+
+The unescaped form is essentially a single namespace byte plus and arbitrary-length sequence of arbitrary bytes, in which a zero byte separates parts.
+The escaped name is a more complicated representation as a zero-terminated string that maps 1:1 onto unescaped names.
+More details can be found in [the relevant docs](/doc/KEYNAMES.md).
+
+There is a conflict between these two forms in terms of API convenience and efficient execution.
+Generally, the escaped form is more convenient to use, since it is entirely human-readable and just a "normal" string.
+When it comes to execution, however, the unescaped name often makes things more efficient.
+For example, order comparisons between unescaped names are much simpler.
```

Comment:
Haha... Now that's the opposite to above. Yes, it's more efficient too. But a single `memcmp` is clearly much simpler than whatever `strchr`/`strncmp` mess you'd have to do with just the escaped name.

Reply:
Huh, can't you `memcmp` a string too?

Reply:
Sure you can, but we want this order:

```c
user:/
user:/a
user:/a/b
user:/a+
```

Because `user:/a < user:/a+` and everything below `user:/a` should be grouped.
But a simple lexicographical sort (i.e. sort by `strncmp`/`memcmp`) gives this order:

```c
user:/
user:/a
user:/a+
user:/a/b
```

`user:/a+ < user:/a/b` because at index 3 we have `+ < /`. With the unescaped name this is not an issue, because the names would be:

```c
\x06\x00
\x06\x00a
\x06\x00a\x00b
\x06\x00a+
```

and `\x00` is the smallest value, it always comes first. Therefore, `\x06\x00a\x00b < \x06\x00a+` since at index 3 we have `\x00 < /`.

> **Note** Yes, `strcmp` on escaped names works for alphanumeric names, and _some_ other symbols, but not all. That's why I used `a+` and not e.g. `ab`.